challenges/2021-08-31-zero-point/solutions/rust/ericwburden/ericwburden.rs

```diff
@@ -0,0 +1,249 @@
+        let mut cur_link = self.head.take();
+        while let Some(mut boxed_node) = cur_link {
+            cur_link = boxed_node.next.take();
+        }
+    }
+}
+
+
+//------------------------------------------------------------------------------------- 
+// Here we create a PushValue trait and implement it for the LinkedList and Links
+//------------------------------------------------------------------------------------- 
+pub trait PushValue<T> {
+    //! This trait allows for pushing values onto the end of the Linked List
+    fn push_value(&mut self, value: T);
+}
+
+/// The `Link`s do all the heavy lifting. If a `Link` is `None`, then it replaces
+/// itself with a Link containing the pushed value. If the `Link` is already populated,
+/// then it pushes the value to it's `next` Link.
+impl<T> PushValue<T> for Link<T> {
+    fn push_value(&mut self, value: T) {
+        if let Some(n) = self {
+            n.next.push_value(value);
+        } else {
+            let new_node = Box::new(Node { value, next: None });
+            *self = Some(new_node);
+        }
+    }
+}
+
+/// The LinkedList struct just passes the value to its `head`, and thence on down the
+/// chain of Links. If 'head' is missing, use the value as a new `head`.
+impl<T> PushValue<T> for LinkedList<T> {
+    fn push_value(&mut self, value: T) {
+        if self.head.is_none() {
+            let new_node = Box::new(Node { value, next: None });
+            self.head = Some(new_node);
+        } else {
+            self.head.push_value(value);
+        }
+    }
+}
+
+
+//------------------------------------------------------------------------------------- 
+// Let's also implement a trait that allows us to push `Link`s to the Linked List,
+// to facilitate moving `Link`s around directly
+//------------------------------------------------------------------------------------- 
+
+pub trait PushLink<T> {
+    fn push_link(&mut self, link: Link<T>);
+}
+
+impl<T> PushLink<T> for Link<T> {
+    fn push_link(&mut self, link: Link<T>) {
+        if let Some(node) = self {
+            node.next.push_link(link);
+        } else {
+            *self = link;
+        }
+    }
+}
+
+impl<T> PushLink<T> for LinkedList<T> {
+    fn push_link(&mut self, link: Link<T>) {
+        if self.head.is_none() {
+            self.head = link;
+        } else {
+            self.head.push_link(link);
+        }
+    }
+}
+
+
+//------------------------------------------------------------------------------------- 
+// Here we implement iteration for our LinkedList (yep, have to do that manually)
+//------------------------------------------------------------------------------------- 
+
+pub struct LinkedListIter<'a, T> {
+    next: Option<&'a Node<T>>,
+}
+
+impl<'a, T> Iterator for LinkedListIter<'a, T> {
+    type Item = &'a Node<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.map(|node| {
+            self.next = node.next.as_deref();
+            node
```

Comment:
here's another example where you just state an object at the end of a function. Curious as to what this does in rust.

Reply:
It's returning that value. I should probably be explicit about returns when sharing code with other folks who are probably more used to using the `return` keyword every time. The upside to this is that you can do things like `let x = if a > b { a } else { b }`, since the `if` statement *also* returns the value of the last unterminated expression.